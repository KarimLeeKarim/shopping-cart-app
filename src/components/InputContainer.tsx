import React from "react";

export const InputContainer: React.FC<InputContainerType> = ({
  formData,
  handleInputData,
  addNewItem,
}) => {    
  return (
    <form onSubmit={addNewItem} className="add-item">
      <input
        type="text"
        id="name"
        placeholder="Name"
        value={formData.name}
        onChange={handleInputData}
        pattern="^[a-zA-Z]+$"
      />
      <input
        type="number"
        id="price"
        placeholder="Price"
        value={formData.price}
        onChange={handleInputData}
        min="1"
      />
      <input
        type="number"
        id="quantity"
        placeholder="Quantity"
        value={formData.quantity}
        onChange={handleInputData}
        min="1"
      />
      <button
        className="submit-form"
        type="submit"
        disabled={!formData.name || !formData.price || !formData.quantity}
        style={{
          background:
            !formData.name || !formData.price || !formData.quantity
              ? "rgb(141, 108, 202)"
              : "rgb(70, 19, 165)",
          cursor:
            !formData.name || !formData.price || !formData.quantity
              ? "auto"
              : "pointer",
        }}
      >
        Add item
      </button>
    </form>
  );
};
