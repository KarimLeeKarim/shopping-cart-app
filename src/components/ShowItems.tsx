import * as React from "react";
import { Dispatch } from "redux";
import { shallowEqual, useDispatch, useSelector } from "react-redux";
import {
  addQuantityItem,
  decreaseQuantityItem,
  removeItem,
  summaryItems,
} from "../redux/actions/ActionCreator";
import { v4 as uuidv4 } from "uuid";
import { dataText } from "../configs/config";

export const ShowItems: React.FC = () => {
  const dispatch: Dispatch<any> = useDispatch();

  const items = useSelector((state: DataState) => {
    return state?.items;
  }, shallowEqual);

  const deleteItem = (item: DataItem) => {
    dispatch(removeItem(item));
    dispatch(summaryItems());
  };

  const addQuantity = (plusQuantityItem: DataItem) => {
    dispatch(addQuantityItem(plusQuantityItem));
    dispatch(summaryItems());
  };

  const decreaseQuantity = (minusQuantityItem: DataItem) => {
    dispatch(decreaseQuantityItem(minusQuantityItem));
    dispatch(summaryItems());
  };

  return (
    <>
      <table>
        <thead>
          <tr>
            <th colSpan={1} className="thead-name">
              {dataText.name}
            </th>
            <th className="thead-price">{dataText.price}</th>
            <th className="thead-quantity">{dataText.quantity}</th>
            <th className="thead-quantity">{dataText.total}</th>
            <th className="thead-remove">{dataText.remove}</th>
          </tr>
        </thead>
        <tbody>
          {items?.map((item: DataItem) => (
            <tr key={uuidv4()}>
              <td className="tbody-name">{item.name}</td>
              <td className="tbody-price">{item.price} $</td>
              <td className="tbody-quantity">
                <button
                  className="tbody-plus"
                  onClick={() => addQuantity(item)}
                >
                  {dataText.plus}
                </button>
                <p>{item.quantity}</p>
                <button
                  className="tbody-minus"
                  onClick={() => decreaseQuantity(item)}
                  disabled={item.quantity === 1 ? true : false}
                  style={{
                    background: item.quantity === 1 ? "#f8864d" : "#fb5301",
                  }}
                >
                  {dataText.minus}
                </button>
              </td>
              <td className="tbody-total">{item.total} $</td>
              <td>
                <button
                  className="tbody-remove"
                  onClick={() => deleteItem(item)}
                >
                  {dataText.delete}
                </button>
              </td>
            </tr>
          ))}
        </tbody>
      </table>
    </>
  );
};
