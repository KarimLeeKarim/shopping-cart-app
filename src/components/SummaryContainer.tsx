import React from "react";
import { shallowEqual, useDispatch, useSelector } from "react-redux";
import { Dispatch } from "redux";
import { clearItems } from "../redux/actions/ActionCreator";
import { v4 as uuidv4 } from "uuid";
import { dataText } from "../configs/config";

export const SummaryContainer: React.FC = () => {
  const dispatch: Dispatch<any> = useDispatch();

  const items = useSelector((state: DataState) => {
    return state?.items;
  }, shallowEqual);

  const summary = useSelector((state: any) => {
    return state.summary;
  }, shallowEqual);

  const clicked = () => {
    dispatch(clearItems());
  };

  return (
    <>
      <div className="block">
        <h1 className="summary-container__title">{dataText.summary}</h1>
        <div className="summary-container__subtitle">
          <div className="summary-container__subtitle__count">
            <p>{dataText.countProduct}</p>
            <p>{items.length}</p>
          </div>
          <div className="summary-container__subtitle__total">
            <p> {dataText.totalProduct}</p> <p>{summary} $</p>
          </div>
          <hr></hr>
        </div>
      </div>
      <button className="summary-container__checkout-btn" onClick={clicked}>
        {dataText.checkout}
      </button>
    </>
  );
};
