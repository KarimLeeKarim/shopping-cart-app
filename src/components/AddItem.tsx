import React, { useState } from "react";
import { InputContainer } from "./InputContainer";

export const AddItem: React.FC<Props> = ({ saveFormData }) => {
  const [formData, setFormData] = useState<DataItemInput>({
    name: "",
    price: "",
    quantity: "",
  });

  const handleInputData = (e: React.FormEvent<HTMLInputElement>) => {
    setFormData({
      ...formData,
      [e.currentTarget.id]: e.currentTarget.value,
    });
  };

  const addNewItem = (e: React.FormEvent) => {
    e.preventDefault();
    saveFormData(formData);
    setFormData({ name: "", price: "", quantity: "" });
  };

  return (
      <InputContainer formData={formData} handleInputData={handleInputData} addNewItem={addNewItem} />  );
};

export default AddItem;
