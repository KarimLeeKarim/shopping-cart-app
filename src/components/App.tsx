import * as React from "react";
import { useDispatch } from "react-redux";
import { Dispatch } from "redux";
import {
  addItem,
  showItem,
  summaryItems,
} from "../redux/actions/ActionCreator";
import { ShowItems } from "./ShowItems";
import { AddItem } from "./AddItem";
import { SummaryContainer } from "./SummaryContainer";
import { dataText, url } from "../configs/config";
import "../assets/App.css";

const App: React.FC = () => {
  const dispatch: Dispatch<any> = useDispatch();

  React.useEffect(() => {
    var xhr = new XMLHttpRequest();
    xhr.open(url.methodGet, url.data, true);
    xhr.onload = function () {
      if (xhr.status === 200) {
        let json = JSON.parse(xhr.responseText);
        let value = json.articles;
        dispatch(showItem(value));
      } else {
        console.error(xhr.statusText);
      }
    };
    xhr.onerror = function (e) {
      console.error(xhr.statusText);
    };
    xhr.send(null);
  }, [dispatch]);

  const saveFormData = React.useCallback(
    (newData: DataItem) => {
      dispatch(addItem(newData));
      dispatch(summaryItems());
    },
    [dispatch]
  );

  return (
    <main>
      <div className="block-show-summary-container">
        <div className="show-items">
          <h1 className="show-items__title">{dataText.cart}</h1>
          <ShowItems />
        </div>
        <div className="summary-container">
          <SummaryContainer />
        </div>
      </div>
      <AddItem saveFormData={saveFormData} />
    </main>
  );
};

export default App;
