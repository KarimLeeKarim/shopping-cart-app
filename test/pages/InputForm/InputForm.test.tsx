/* eslint-disable @typescript-eslint/no-unused-vars */
import React from "react";
import { InputContainer } from "../../../src/components/InputContainer";
import { Provider } from "react-redux";
import { combineReducers, createStore } from "redux";
import { render, ShallowWrapper } from "enzyme";
import { shallowToJson } from "enzyme-to-json";
import { mount, shallow } from "../../setup/test-setup";

const mockfn = jest.fn();

describe("<AddItem/>", () => {
  let wrapper: any;

  const props: any = {
    formData: mockfn,
    handleInputData: mockfn,
    addNewItem: mockfn,
  };

  beforeEach(() => {
    wrapper = shallow(<InputContainer {...props} />);
  });

  it("should match the snapshot", () => {
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  describe("define fields", () => {
    it("reder product name", () => {
      const productName = wrapper.find("input[id='name']");

      expect(productName.prop("type")).toBe("text");
      expect(productName.prop("id")).toBe("name");
      expect(productName.prop("placeholder")).toBe("Name");
      productName.simulate("change");
    });

    it("reder product price", () => {
      const productName = wrapper.find("input[id='price']");

      expect(productName.prop("type")).toBe("number");
      expect(productName.prop("id")).toBe("price");
      expect(productName.prop("placeholder")).toBe("Price");
      productName.simulate("change");
    });
    it("reder product quantity", () => {
      const productName = wrapper.find("input[id='quantity']");

      expect(productName.prop("type")).toBe("number");
      expect(productName.prop("id")).toBe("quantity");
      expect(productName.prop("placeholder")).toBe("Quantity");
      productName.simulate("change");
    });
    it("render product submit button", () => {
      const btn =wrapper.find("button").at(0);
      expect(btn.prop('type')).toBe('submit');
      expect(btn.text()).toEqual('Add item');
    });
  });
});
