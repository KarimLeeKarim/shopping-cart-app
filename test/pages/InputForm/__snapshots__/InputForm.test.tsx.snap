// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddItem/> should match the snapshot 1`] = `
<form
  className="add-item"
  onSubmit={[MockFunction]}
>
  <input
    id="name"
    onChange={[MockFunction]}
    pattern="^[a-zA-Z]+$"
    placeholder="Name"
    type="text"
    value="mockConstructor"
  />
  <input
    id="price"
    min="1"
    onChange={[MockFunction]}
    placeholder="Price"
    type="number"
  />
  <input
    id="quantity"
    min="1"
    onChange={[MockFunction]}
    placeholder="Quantity"
    type="number"
  />
  <button
    className="submit-form"
    disabled={true}
    style={
      Object {
        "background": "rgb(141, 108, 202)",
        "cursor": "auto",
      }
    }
    type="submit"
  >
    Add item
  </button>
</form>
`;
